File: services/mini-coupang/src/main/java/com/deepblog/minicoupang/domain/product/service/ProductService.java
package com.deepblog.minicoupang.domain.product.service;

import com.deepblog.minicoupang.domain.product.Product;
import com.deepblog.minicoupang.domain.product.api.dto.CreateProductRequest;
import com.deepblog.minicoupang.domain.product.api.dto.ProductResponse;
import com.deepblog.minicoupang.domain.product.storage.ProductRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.server.ResponseStatusException;

import java.util.List;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class ProductService {

    private final ProductRepository repository;

    @Transactional
    public ProductResponse create(CreateProductRequest request) {
        Product product = Product.create(
                request.sellerId(),
                request.sellerName(),
                request.name(),
                request.price(),
                request.stock());
        Product saved = repository.save(product);
        return ProductResponse.from(saved);
    }

    public ProductResponse findById(Long id) {
        Product product = repository.findById(id).orElseThrow(() ->
                new ResponseStatusException(HttpStatus.NOT_FOUND, "Product not found: " + id));
        return ProductResponse.from(product);
    }

    public List<ProductResponse> findBySellerId(Long sellerId) {
        return repository.findBySellerId(sellerId).stream()
                .map(ProductResponse::from)
                .toList();
    }
}

File: services/product-service/src/main/kotlin/com/deepblog/product/service/ProductService.kt
package com.deepblog.product.service

import com.deepblog.product.api.dto.CreateProductRequest
import com.deepblog.product.api.dto.ProductResponse
import com.deepblog.product.domain.Product
import com.deepblog.product.storage.ProductRepository
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import org.springframework.web.server.ResponseStatusException

@Service
@Transactional(readOnly = true)
class ProductService(
    private val repository: ProductRepository,
) {
    @Transactional
    fun create(request: CreateProductRequest): ProductResponse {
        val product = Product(
            sellerId = request.sellerId,
            sellerName = request.sellerName.trim(),
            name = request.name.trim(),
            price = request.price,
            stock = request.stock,
        )
        val saved = repository.save(product)
        return ProductResponse.from(saved)
    }

    fun findById(id: Long): ProductResponse {
        val product = repository.findById(id).orElseThrow {
            ResponseStatusException(HttpStatus.NOT_FOUND, "Product not found: $id")
        }
        return ProductResponse.from(product)
    }

    fun findBySellerId(sellerId: Long): List<ProductResponse> =
        repository.findBySellerId(sellerId).map { ProductResponse.from(it) }
}
